File: scripts/probe_vocos.py
#!/usr/bin/env python3
"""Reference probe for the Vocos vocoder: export its weights as safetensors for
the Burn port, run it on a fixed mel, and dump intermediates for parity.

    python3 scripts/probe_vocos.py
"""
import os

import numpy as np
import torch
from safetensors.torch import save_file
from vocos import Vocos

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
W = os.path.join(ROOT, "weights")
OUT = os.path.join(ROOT, "probes", "vocos")
os.makedirs(W, exist_ok=True)
os.makedirs(OUT, exist_ok=True)

torch.manual_seed(0)
torch.set_grad_enabled(False)

v = Vocos.from_pretrained("charactr/vocos-mel-24khz").eval()
save_file({k: t.clone().contiguous() for k, t in v.state_dict().items()}, os.path.join(W, "vocos.safetensors"))
print("exported weights → weights/vocos.safetensors")

T = 64
mel = torch.randn(1, 100, T)
np.save(os.path.join(OUT, "mel.npy"), mel.numpy().astype(np.float32))

probes = {}

def hook(name):
    def f(_m, _i, o):
        probes[name] = (o[0] if isinstance(o, tuple) else o).detach().numpy().astype(np.float32)
    return f

v.backbone.embed.register_forward_hook(hook("embed"))
v.backbone.final_layer_norm.register_forward_hook(hook("backbone"))
v.head.out.register_forward_hook(hook("head_out"))

audio = v.decode(mel)  # backbone → head (ISTFT)
probes["audio"] = audio.detach().numpy().astype(np.float32)

for k, val in probes.items():
    np.save(os.path.join(OUT, k + ".npy"), val)
    print(f"{k}: {tuple(val.shape)}")
print(f"audio len {audio.shape[-1]} = (T-1)*hop = {(T-1)*256}  range [{audio.min():.3f},{audio.max():.3f}]")

File: src/bin/probe_vocos.rs
//! Rust side of the Vocos numerical-parity probe. Loads the fixed mel from
//! probes/vocos/mel.npy and the exported weights, runs `Vocos::forward_probed`,
//! and writes intermediates to probes/vocos_rust/ for compare_vocos.py.
//!
//!   python3 scripts/probe_vocos.py            # reference + weight export
//!   cargo run --release --bin probe_vocos

use mary::nn::backend::B;
use mary::nn::npy;
use mary::nn::weight_loader::{SingleFileLoader, WeightLoader};
use burn::prelude::*;
use burn::tensor::TensorData;
use std::path::{Path, PathBuf};
use mary::models::f5::vocos::Vocos;

fn main() {
    let device = Default::default();
    let root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let py = root.join("probes").join("vocos");
    let rs = root.join("probes").join("vocos_rust");
    std::fs::create_dir_all(&rs).unwrap();

    let (md, ms) = npy::load_npy(&py.join("mel.npy")).unwrap();
    let mel = Tensor::<B, 3>::from_data(TensorData::new(md, ms), &device);

    let wpath = root.join("weights").join("vocos.safetensors");
    let loader = WeightLoader::SingleFile(SingleFileLoader::new(&wpath));
    let model = Vocos::<B>::load(&loader, &device);

    let (_audio, probes) = model.forward_probed(mel);
    for (name, data, shape) in &probes {
        npy::save_npy(&rs.join(format!("{name}.npy")), data, shape).unwrap();
        println!("{name}: {shape:?}");
    }
    println!("✓ vocos rust probes → {}", rs.display());
}
